File: src/main/java/eu/karelhovorka/zpevnik/formatter/MustacheTemplateSongFormatter.java
package eu.karelhovorka.zpevnik.formatter;


import eu.karelhovorka.zpevnik.text.SongText;
import eu.karelhovorka.zpevnik.util.Transposer;

public abstract class MustacheTemplateSongFormatter extends SongFormatter {

    protected final String mainTemplate;

    protected final String chordReplaceTemplate;

    protected SongText context;

    protected MustacheTemplateSongFormatter(String mainTemplate, String chordReplaceTemplate) {
        this.mainTemplate = mainTemplate;
        this.chordReplaceTemplate = chordReplaceTemplate;
    }

    public String formatHtml(SongText context) {
        this.context = context;
        return formatHtmlChords(applyTemplate(mainTemplate, context));
    }

    protected abstract String applyTemplate(String template, SongText context);

    public String formatHtmlChords(String content) {
        return content.replaceAll(Transposer.CHORD_REGEX, applyTemplate(chordReplaceTemplate, context));
    }

}

File: src/main/java/eu/karelhovorka/zpevnik/formatter/MustacheTemplateSongFormatter.kt
package eu.karelhovorka.zpevnik.formatter


import eu.karelhovorka.zpevnik.text.SongText
import eu.karelhovorka.zpevnik.util.Transposer

abstract class MustacheTemplateSongFormatter protected constructor(protected val mainTemplate: String, protected val chordReplaceTemplate: String) : SongFormatter() {

    protected lateinit var songText: SongText

    override fun formatHtml(songText: SongText): String {
        this.songText = songText
        return formatHtmlChords(applyTemplate(mainTemplate, songText))
    }

    protected abstract fun applyTemplate(template: String, context: SongText): String

    public override fun formatHtmlChords(content: String): String {
        return content.replace(Transposer.CHORD_REGEX.toRegex(), applyTemplate(chordReplaceTemplate, songText))
    }

}
